### PR TITLE
Link mesh files for visual and collision instead of creating indexed face sets

### PR DESCRIPTION
**Pull request by BenjaminHug8**
This PR is for linking the mesh files instead of transforming them into index face sets when converting URDF to PROTO, thus the PROTO files are way more shorter and Webots can load faster.

### CODE REVIEW

**Comment by BenjaminHug8** [comment]
There is code in commentary that I will delete once you approved (maybe should we keep it) ?

**Comment by BenjaminHug8** [comment]
I just noticed one thing in Human.proto, the closing `}` is not correctly indented, I am on it.

**Comment by omichel** [comment]
> There is code in commentary that I will delete once you approved (maybe should we keep it) ?

Yes, we should delete it.

**Comment by BenjaminHug8** [comment]
Ok I removed the code in commentaries and check once the code, so this PR should be ok now.

I let you @omichel do a final check with the review.

File: urdf2webots/parserURDF.py [hunk 1]
```diff
@@ -87,6 +88,14 @@ def __init__(self):
         self.radius = 0.0
 
 
+class Mesh():
+    """Define mesh object."""
+
+    def __init__(self):
+        """Initializatization."""
+        self.url = ""
```
Comment (omichel) [inline urdf2webots/parserURDF.py:96]: We prefer single quotes:
```suggestion
        self.url = ''
```

File: urdf2webots/parserURDF.py [hunk 2]
```diff
@@ -848,17 +642,12 @@ def getCollision(link, node, path):
...
-                getOBJMesh(meshfile, collision, link)
-            elif extension == '.stl':
+            if extension in extensionList:
                 name = os.path.splitext(os.path.basename(meshfile))[0]
                 if name in Geometry.reference:
                     collision.geometry = Geometry.reference[name]
                 else:
-                    if extension == '.stl':
-                        collision.geometry.stl = getSTLMesh(meshfile, collision)
+                    collision.geometry.mesh.url = "\"" + meshfile + "\""
```
Comment (omichel) [inline urdf2webots/parserURDF.py:650]:
```suggestion
                    collision.geometry.mesh.url = '"' + meshfile + '"'
```

**Review by omichel — CHANGES_REQUESTED** [review]
It looks good to me. Only a few minor things.

File: urdf2webots/parserURDF.py [hunk 3]
```diff
@@ -786,17 +585,12 @@ def getVisual(link, node, path):
...
-                getOBJMesh(meshfile, visual, link)
-            elif extension == '.stl':
+            if extension in extensionList:
                 name = os.path.splitext(os.path.basename(meshfile))[0]
                 if name in Geometry.reference:
                     visual.geometry = Geometry.reference[name]
                 else:
-                    if extension == '.stl':
-                        visual = getSTLMesh(meshfile, visual)
+                    visual.geometry.mesh.url = "\"" + meshfile + "\""
```
Comment (omichel) [inline urdf2webots/parserURDF.py:593]: Missed that one:
```suggestion
                    visual.geometry.mesh.url = '"' + meshfile + '"'
```

File: urdf2webots/writeProto.py [hunk 4]
```diff
@@ -397,93 +376,20 @@ def URDFVisual(proto, visualNode, level, normal=False):
...
-                        proto.write('%d %d %d -1 ' % (value[0], value[1], value[2]))
-                    elif len(value) == 2:
-                        assert visualNode.geometry.lineset
-                        proto.write('%d %d -1 ' % (value[0], value[1]))
-            elif isinstance(visualNode.geometry.trimesh.coordIndex[0], np.int32):
-                for i in range(int(len(visualNode.geometry.trimesh.coordIndex) / 3)):
-                    proto.write('%d %d %d -1 ' % (visualNode.geometry.trimesh.coordIndex[3 * i + 0],
-                                                  visualNode.geometry.trimesh.coordIndex[3 * i + 1],
-                                                  visualNode.geometry.trimesh.coordIndex[3 * i + 2]))
+                proto.write((shapeLevel + 1) * indent + 'geometry DEF %s Mesh {\n' % (visualNode.geometry.defName))
```
Comment (omichel) [inline urdf2webots/writeProto.py:386]: I believe these extra parenthesis are not needed:
```suggestion
                proto.write((shapeLevel + 1) * indent + 'geometry DEF %s Mesh {\n' % visualNode.geometry.defName)
```

**Review by omichel — CHANGES_REQUESTED** [review]

**Comment by omichel** [comment]
After this change, it seems that the following options are now obsolete: `--disable-mesh-optimization` `--multi-file`. We should probably remove them and update the Readme.md accordingly.

**Comment by BenjaminHug8** [comment]
> After this change, it seems that the following options are now obsolete: `--disable-mesh-optimization` `--multi-file`. We should probably remove them and update the Readme.md accordingly.

I also remove `Trimesh` as it was used before for the conversion of mesh into index face set.

File: urdf2webots/writeProto.py [hunk 5]
```diff
@@ -7,7 +7,6 @@
 
 toolSlot = None
 staticBase = False
-enableMultiFile = False
 meshFilesPath = None
```
Comment (omichel) [inline urdf2webots/writeProto.py:10]: This one seems to be useless now:
```suggestion
```

**Review by omichel — CHANGES_REQUESTED** [review]

**Comment by BenjaminHug8** [comment]
Should I update the `Readme.md` to state that we removed the 2 options and that now we are directly linking the meshes instead or is it not necessary ?

**Comment by omichel** [comment]
I believe we should introduce a CHANGELOG.md file at the top level and start logging this kind of changes under version numbers.

**Comment by omichel** [comment]
> I believe we should introduce a CHANGELOG.md file at the top level and start logging this kind of changes under version numbers.

No, that's not a good idea as it seems there is already some automatic generation of change log, see the [releases page](https://github.com/cyberbotics/urdf2webots/releases).
So we should simply ensure this automatic change log contains clear information about the change (thus the title of this PR is probably super important).

**Comment by BenjaminHug8** [comment]
> > I believe we should introduce a CHANGELOG.md file at the top level and start logging this kind of changes under version numbers.
> 
> No, that's not a good idea as it seems there is already some automatic generation of change log, see the [releases page](https://github.com/cyberbotics/urdf2webots/releases). So we should simply ensure this automatic change log contains clear information about the change (thus the title of this PR is probably super important).

So can we just have to merge this PR and prepare for a new release ?
I am just wondering why we get from version `1.0.9` to `10.0.10` ?

**Comment by omichel** [comment]
> So can we just have to merge this PR and prepare for a new release?

Yes. That's probably a new GitHub feature that was not available before 10.0.10 or that we didn't enable when making the previous releases.

**Review by omichel — APPROVED** [review]
Thank you.

**Pull request merged**